src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala

```diff
@@ -49,15 +49,98 @@ object JndiSchemaDAO {
   }
 }
 
-class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport {
+class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig, val schemaLockConfig: SchemaLockConfig)(implicit executionContext: ExecutionContext) extends JndiSupport with DirectorySubjectNameSupport with LazyLogging {
+
+  def initSchemaLock(): Future[Unit] = {
+    for {
+      _ <- createSchemaChangeLockSchema()
+      _ <- createOrgUnit(schemaLockOu)
+      _ <- insertSchemaLock()
+    } yield ()
+  }
+
+  def cleanupSchemaLock(): Future[Unit] = {
+    for {
+      _ <- removeSchemaLock()
+      _ <- removeSchemaChangeLockSchema()
+    } yield ()
+  }
 
   def init(): Future[Unit] = {
     for {
+      _ <- if(schemaLockConfig.lockSchemaOnBoot) initSchemaLock() else Future.successful(())
       _ <- destroySchema()
       _ <- createSchema()
+      _ <- if(schemaLockConfig.lockSchemaOnBoot) cleanupSchemaLock() else Future.successful(())
     } yield ()
   }
 
+  def insertSchemaLock(): Future[Unit] = withContext { ctx =>
+    logger.info("Acquiring schema lock...")
+
+    val resourceContext = new BaseDirContext {
+      override def getAttributes(name: String): Attributes = {
+        val myAttrs = new BasicAttributes(true) // Case ignore
+
+        val oc = new BasicAttribute("objectclass")
+        Seq("top", "lockId").foreach(oc.add)
+        myAttrs.put(oc)
+        myAttrs.put("lockId", "0")
+        myAttrs.put("instanceId", schemaLockConfig.instanceId)
+
+        myAttrs
+      }
+    }
+
+    try {
+      ctx.bind(schemaLockDn, resourceContext)
+    } catch {
+      case _: NameAlreadyBoundException => {
+        logger.info("Schema is already locked. Waiting for schema lock...")
+        waitForSchemaLock(schemaLockConfig.maxTimeToWait / schemaLockConfig.recheckTimeInterval)
+      }
+    }
+  }
+
+  def removeSchemaLock(): Future[Unit] = withContext { ctx =>
+    logger.info("Releasing schema lock...")
+
+    ctx.unbind(schemaLockDn)
+  }
+
+  def checkSchemaLock(): Future[Boolean] = withContext { ctx =>
+    val attributes = Try { ctx.getAttributes(schemaLockDn) }
+
+    attributes match {
+      case Success(attrs) => {
+        val instanceId = attrs.get("instanceId").get.toString
+        logger.info(s"Schema currently locked by sam instance [$instanceId]...")
+        true
+      }
+      case Failure(_) => false
+    }
+  }
+
+  def waitForSchemaLock(remainingAttempts: Int): Future[Unit] = withContext { ctx =>
```

Comment:
I don't think you need a checkSchemaLock function. All this can be a couple lines of code in insertSchemaLock that ends with a recursive call to insertSchemaLock: How about insertSchemaLock takes the number of times to try and interval as parameters. Then attempt insert, catch exception, log message, sleep for interval, call insertSchemaLock decrementing count.